src/api/app/controllers/source_project_package_meta_controller.rb

```diff
@@ -0,0 +1,86 @@
+class SourceProjectPackageMetaController < SourceController
+  # override the ApplicationController version
+  # to have meaningful error messages
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+  before_action :require_package_name, only: [:show, :update]
+  before_action :set_rdata, only: [:update]
+  before_action :validate_project_name, only: [:update]
+  before_action :validate_package_name, only: [:update]
+  validate_action update: { request: :package, response: :status }
+
+  # GET /source/:project/:package/_meta
+  def show
+    pack = Package.get_by_project_and_name(@project_name, @package_name, use_source: false)
+
+    if params.key?(:meta) || params.key?(:rev) || params.key?(:view) || pack.nil?
+      # check if this comes from a remote project, also true for _project package
+      # or if meta is specified we need to fetch the meta from the backend
+      path = request.path_info
+      path += build_query_from_hash(params, [:meta, :rev, :view])
+      pass_to_backend path
+      return
+    end
+
+    render xml: pack.to_axml
+  end
+
+  # PUT /source/:project/:package/_meta
+  def update
+    # check for project
+    if Package.exists_by_project_and_name(@project_name, @package_name, follow_project_links: false)
+      pkg = Package.get_by_project_and_name(@project_name, @package_name, use_source: false)
+
+      authorize pkg, :update?
+
+      return if change_package_protection_level?(pkg)
+    else
+      prj = Project.get_by_name(@project_name)
+      # necessary to pass the policy_class here
+      # if its remote prj is a string
+      authorize prj, :update?, policy_class: ProjectPolicy
+      pkg = prj.packages.new(name: @package_name)
+    end
+
+    pkg.set_comment(params[:comment])
+    pkg.update_from_xml(@rdata)
+    render_ok
+  end
+
+  private
+
+  def change_package_protection_level?(pkg)
+    # TODO: use pundit
+    if pkg && !pkg.disabled_for?('sourceaccess', nil, nil)
+      if FlagHelper.xml_disabled_for?(@rdata, 'sourceaccess') && !User.current.is_admin?
+        render_error status: 403, errorcode: 'change_package_protection_level',
+                     message: 'admin rights are required to raise the protection level of a package'
+        return true
+      end
+    end
+    false
+  end
+
+  def require_package_name
+    required_parameters :project, :package
+
+    @project_name = params[:project]
+    @package_name = params[:package]
+
+    valid_package_name! @package_name
+  end
+
+  def validate_package_name
+    render_error_for_package_or_project 'package_name_mismatch',
+                                        'package name in xml data does not match resource path component',
```

Comment:
actually the message could be reuse as well... I mean it is just the same message changing package by project... :thinking: